aten/src/ATen/native/mkldnn/Conv.cpp

```diff
@@ -884,6 +885,84 @@ std::tuple<Tensor, Tensor, Tensor> mkldnn_convolution_backward(
 
 REGISTER_ALL_CPU_DISPATCH(mkldnn_convolution_backward_stub, &mkldnn_convolution_backward);
 
+namespace functionalization {
+
+Tensor& mkldnn_convolution_pointwise_binary_(
+    const Tensor& input_t,
+    Tensor& other_t,
+    const Tensor& weight_t,
+    const c10::optional<Tensor>& bias_opt,
+    IntArrayRef padding,
+    IntArrayRef stride,
+    IntArrayRef dilation,
+    int64_t groups,
+    c10::string_view binary_attr,
+    c10::optional<at::Scalar> alpha,
+    c10::optional<c10::string_view> unary_attr,
+    torch::List<c10::optional<at::Scalar>> unary_scalars,
+    c10::optional<c10::string_view> unary_algorithm) {
+  // other_t += convolution(...), other_t = unary(other_t)
+  TORCH_CHECK(
+      input_t.ndimension() == 4 || input_t.ndimension() == 5,
+      "functional mkldnn_convolution_pointwise_binary_: currently only support 2d and 3d")
+  TORCH_CHECK(
+      binary_attr == "add",
+      "mkldnn_convolution_pointwise_binary_: only support binary op fusion")
+  TORCH_CHECK(
+      !alpha.has_value() || alpha.value().to<float>() == 1.0,
+      "functional mkldnn_convolution_pointwise_binary_: the alpha value for the binary op should be none(meaning 1.0) or 1.0");
+  TORCH_CHECK(
+      !unary_attr.has_value() || unary_attr.value() == "relu",
+      "functional mkldnn_convolution_pointwise_binary_: only support none or relu unary op fusion after binary op");
+  bool is_functional_inputs = true;
+  is_functional_inputs &= at::functionalization::impl::isFunctionalTensor(input_t);
+  is_functional_inputs &= at::functionalization::impl::isFunctionalTensor(other_t);
+  is_functional_inputs &= at::functionalization::impl::isFunctionalTensor(weight_t);
+  is_functional_inputs &= at::functionalization::impl::isFunctionalTensor(bias_opt);
+
+  TORCH_CHECK(
+      is_functional_inputs,
+      "functionalization mkldnn_convolution_pointwise_binary_: the all inputs should be FunctionalTensor")
+   at::Tensor input_;
+  at::functionalization::impl::sync(input_t);
+  input_ = at::functionalization::impl::from_functional_tensor(input_t);
+
+  at::Tensor other_;
+  at::functionalization::impl::sync(other_t);
+  other_ = at::functionalization::impl::from_functional_tensor(other_t);
+
+  at::Tensor weight_;
+  at::functionalization::impl::sync(weight_t);
+  weight_ = at::functionalization::impl::from_functional_tensor(weight_t);
+
+  c10::optional<at::Tensor> bias_;
+  at::functionalization::impl::sync(bias_opt);
+  bias_ = at::functionalization::impl::from_functional_tensor(bias_opt);
+
+  at::AutoDispatchSkipFunctionalize guard;
+  Tensor conv_output, tmp_output;
+  if (weight_.is_mkldnn()) {
+    conv_output = at::native::mkldnn_convolution(
+        input_, weight_, bias_, padding, stride, dilation, groups);
+  } else {
+    conv_output = at::convolution(
+        input_, weight_, bias_, stride, padding, dilation, false, 0, groups);
+  }
+
+  if (unary_attr.has_value()) {
+    tmp_output = other_.add(conv_output).relu();
```

Comment:
For the functionalization of such an in-place op, we just want to make sure it can work when collecting the meta info, and that it will lower the fused kernel at the inductor path.

Reply:
I see. In that case - if you're okay with decomposing the op (and relying on fusion for perf), then I think there's a simpler alternative here, that doesn't require a custom functionalization kernel. You can write a python decomp for this op, that just decomposes it into `convolution` + inplace add + inplace unary op. And then you can rely on functionalization's existing support for the inplace add + unary ops. The easiest way to write this decomp is in python, registering it to the autograd key (so it decomposes before functionalization). Example [here](https://github.com/pytorch/pytorch/blob/6ac0198c0228a5b59f337cac5a752dbb28e341e5/torch/_decomp/decompositions.py#L1435) (although it's a little unclear to me if the decomp should live directly in that file - does that mkldnn operator always exist in the dispatcher? `mkldnn/Conv.cpp` only get built + that op registered on a build with mkldnn).

Reply:
> and that it will lower the fused kernel at the inductor path.

I might have misunderstood. This function as written won't end up lowering the fused kernel - It will directly lower `mkldnn_convolution` + `aten.add` + `aten.relu` to inductor

Reply:
@bdhirsh , for my understanding, the functionalization is used in https://github.com/pytorch/pytorch/blob/6ac0198c0228a5b59f337cac5a752dbb28e341e5/torch/_functorch/aot_autograd.py#L1378 to collect the meta info, but we don't want to do the decomposing, so for inductor side, we still get the in -place fusion op, and lowering same the in-place kernel:https://github.com/pytorch/pytorch/blob/ca63040d2b7583ae7c8f066fff302473b591f194/torch/_inductor/lowering.py#L902.

Reply:
Yep, this op is not user-generated but a fusion op generated from inside the inductor backend. The reason why we have to explicitly implement a functionalization is that the fusion path is currently added before AOTAutograd. We have the plan to move the fusions after AOTAutograd later so that this functionalization can be removed after the move is completed.

Reply:
> 

I don't think it will lower tto **mkldnn_convolution** + **aten.add** + **aten.relu** to inductor, for the inductor path, if we don't register a decomposing at https://github.com/pytorch/pytorch/blob/6c80d0a5a55de71d0cc56950f7459b5c334a14ab/torch/_decomp/__init__.py#L175, we can't call in the decomposition path even we do https://github.com/pytorch/pytorch/blob/6ac0198c0228a5b59f337cac5a752dbb28e341e5/torch/_decomp/decompositions.py, is it right?

Reply:
Hmm.. so there are a few ways for ops to get "decomposed" in AOTAutograd. One is through the explicit decomp table, like you said. But separately from the decomp table, both autograd and functionalization run inside of AOTAutograd. If either of them chooses to directly decompose an operator, then the compiler sitting underneath AOTAutograd will see the result of that decomposition too.

You can probably try writing a simple test to confirm whether or not inductor will see the fused op:
```
@torch.compile
def f(inp, other, weight):
    # added a clone so that the mutation is not a "graph input mutation"
    other_clone = other.clone()
    return torch.ops.mkldnn.mkldnn_convolution_pointwise_binary_(inp, other, weight)



inp = torch.ones(64)
other = torch.ones(64)
weight = torch.ones(64, requires_grad=True)
out = f(inp, other, weight)
```


FWIW - AOTAutograd has two code paths: `aot_dispatch_autograd` and `aot_dispatch_base`, and only the first actually runs functionalization - so if you are going the "inference path", then you won't see your functionalization kernel actually affect the graph sent to inductor (this happens if none of the inputs to the graph require grad). But before the 2.0 branch cut, I'm planning on getting functionalization to run in both paths of AOTAutograd.